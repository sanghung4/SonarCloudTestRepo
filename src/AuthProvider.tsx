import React, {
  ReactNode,
  createContext,
  useMemo,
  useEffect,
  useState,
  useCallback
} from 'react';

import {
  AuthState,
  OktaAuth,
  SigninWithCredentialsOptions,
  UserClaims
} from '@okta/okta-auth-js';
import { useOktaAuth } from '@okta/okta-react';
import { useHistory, useLocation } from 'react-router-dom';

import ApolloClient from 'client';
import { identify } from 'utils/analytics';
import {
  ApprovedUser,
  Feature,
  Maybe,
  useFeaturesQuery,
  useSetFeatureEnabledMutation,
  useUserQuery
} from 'generated/graphql';
import { selectedAccountsVar } from 'providers/SelectedAccountsProvider';

export type Profile = {
  permissions: Array<string>;
  userId: string;
  isEmployee: boolean;
  isVerified: boolean;
};

export type UserContextType = {
  activeFeatures?: string[];
  authState: AuthState | null;
  ecommUser?: Maybe<ApprovedUser>;
  features?: Feature[];
  featuresLoading?: boolean;
  firstName?: string;
  handleLogout?: (redirectTo?: string) => void;
  isLoggingOut?: boolean;
  lastName?: string;
  login?: (values: SigninWithCredentialsOptions) => Promise<void>;
  oktaAuth?: OktaAuth;
  profile?: Profile;
  setFeature?: (featureId: string, isEnabled: boolean) => void;
  user?: UserClaims;
};

type Props = {
  children: ReactNode;
};

export const AuthContext = createContext<UserContextType>({
  authState: null,
  isLoggingOut: false
});

function AuthProvider(props: Props) {
  /**
   * Custom hooks
   */
  const history = useHistory();
  const location = useLocation();
  const { oktaAuth, authState } = useOktaAuth();

  /**
   * State
   */
  const [user, setUser] = useState<UserClaims>();
  const [profile, setProfile] = useState<Profile>({
    permissions: [] as Array<string>,
    userId: '',
    isEmployee: false,
    isVerified: false
  });
  const [features, setFeatures] = useState<any[]>([]);
  const [isLoggingOut, setIsLoggingOut] = useState(false);

  /**
   * Data
   */
  const { data: featuresQueryData, loading: featuresLoading } =
    useFeaturesQuery();
  const { data: userData } = useUserQuery({
    skip: !profile?.userId,
    variables: {
      userId: profile?.userId
    }
  });

  /**
   * Memo
   */
  const firstName = useMemo(firstNameMemo, [user]);
  const lastName = useMemo(lastNameMemo, [user]);
  const featureData: any = useMemo(featuresMemo, [featuresQueryData?.features]);
  const activeFeatures = useMemo(activeFeaturesMemo, [featuresQueryData]);

  /**
   * Effects
   */
  useEffect(fetchUser, [authState?.isAuthenticated, isLoggingOut, oktaAuth]);
  useEffect(fetchProfile, [
    authState?.accessToken,
    authState?.isAuthenticated,
    isLoggingOut,
    oktaAuth
  ]);
  useEffect(fetchFeatures, [
    featureData,
    authState?.isAuthenticated,
    featuresQueryData?.features
  ]);
  useEffect(() => {
    if (location.pathname === '/' && isLoggingOut) setIsLoggingOut(false);
  }, [isLoggingOut, location.pathname]);
  useEffect(checkVerifiedUser, [history, location, profile]);

  const [setFeatureEnabledMutation] = useSetFeatureEnabledMutation();
  /**
   * Callbacks
   */
  const login = useCallback(loginCb, [oktaAuth]);

  async function loginCb(values: SigninWithCredentialsOptions) {
    const transaction = await oktaAuth.signInWithCredentials(values);
    identify(transaction.user?.profile?.login);
    oktaAuth.signInWithRedirect({
      sessionToken: transaction.sessionToken
    });
  }

  return (
    <AuthContext.Provider
      value={{
        activeFeatures,
        authState,
        ecommUser: userData?.user,
        features,
        featuresLoading,
        firstName,
        handleLogout,
        isLoggingOut,
        lastName,
        login,
        oktaAuth,
        profile,
        setFeature,
        user
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );

  function firstNameMemo() {
    return user?.name?.split(' ')[0] ?? '';
  }
  function lastNameMemo() {
    return user?.name?.split(' ')[1] ?? '';
  }

  function fetchUser() {
    if (authState?.isAuthenticated && !isLoggingOut) {
      const getUser = async () => {
        try {
          const oktaUserData = await oktaAuth.getUser();
          setUser(oktaUserData);
        } catch {}
      };

      getUser();
    }
  }

  function fetchProfile() {
    if (authState?.isAuthenticated && !isLoggingOut) {
      const getProfile = () => {
        const decoded = JSON.parse(
          atob(authState?.accessToken?.accessToken.split('.')[1] ?? '')
        );
        setProfile({
          permissions: decoded.ecommPermissions,
          userId: decoded.ecommUserId,
          isEmployee: decoded.isEmployee,
          isVerified: decoded.isVerified
        });

        if (
          (localStorage.getItem('currentUserId') || '') !== decoded.ecommUserId
        ) {
          localStorage.removeItem('defaultAccounts');
        }

        localStorage.setItem('currentUserId', decoded.ecommUserId);
      };

      getProfile();
    }
  }

  function fetchFeatures() {
    setFeatures(featureData);
  }

  function checkVerifiedUser() {
    if (
      profile.isEmployee &&
      !profile.isVerified &&
      location.pathname !== '/'
    ) {
      history.replace('/');
    }
  }

  function setFeature(featureId: string, isEnabled: boolean) {
    setFeatureEnabledMutation({
      variables: {
        featureId,
        setFeatureEnabledInput: {
          isEnabled
        }
      },
      refetchQueries: ['features']
    });
  }

  async function handleLogout(redirectTo: string = '/') {
    if (authState?.isAuthenticated && !isLoggingOut) {
      setIsLoggingOut(true);
      await oktaAuth.signOut({
        postLogoutRedirectUri: `${window.location.origin}${redirectTo}`
      });

      selectedAccountsVar({});
      localStorage.setItem('selectedAccounts', JSON.stringify({}));
      sessionStorage.clear();
      await ApolloClient.clearStore();
    }
  }

  function featuresMemo() {
    const features = featuresQueryData?.features?.map((feature: any) => {
      return feature;
    });
    return features;
  }

  function activeFeaturesMemo() {
    return (featuresQueryData?.features ?? [])
      .filter((f) => f.isEnabled)
      .map((f) => f.name);
  }
}

export default AuthProvider;
