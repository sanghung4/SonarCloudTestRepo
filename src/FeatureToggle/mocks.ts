import { Feature } from 'generated/graphql';

export const mockFeatures: Feature[] = [
  { id: '1', isEnabled: false, name: 'noooo' },
  { id: '2', isEnabled: true, name: 'fine' },
  { id: '3', isEnabled: true, name: 'test' }
];
