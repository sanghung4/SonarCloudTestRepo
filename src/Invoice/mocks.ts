import { GetInvoiceDocument } from 'generated/graphql';

export const successMincron = {
  request: {
    query: GetInvoiceDocument,
    variables: {
      accountId: '123456',
      invoiceNumber: '5460026'
    }
  },
  result: {
    data: {
      invoice: {
        invoiceNumber: '5460026',
        status: 'Open',
        terms: 'NET 30 DAYS',
        customerPo: '40251/13325',
        invoiceDate: '10/28/2020',
        dueDate: '10/28/2020',
        originalAmt: '',
        openBalance: '',
        age: '',
        jobNumber: 'AVLNEP7',
        jobName: 'NEP 7 ASHEVILLE',
        address: {
          streetLineOne: '911 BLUEBERRY WAY',
          streetLineTwo: '',
          streetLineThree: '',
          city: 'NORTHLAKE',
          state: 'TX',
          postalCode: '76226',
          country: ''
        },
        shipDate: '11/06/2020',
        deliveryMethod: 'Our Truck',
        subtotal: '3184.28',
        tax: '262.76',
        otherCharges: '0.0',
        paidToDate: '1000.00',
        invoiceItems: [
          {
            id: '1738433F4-91A7-4637-8C60-B4DF59A060C1',
            brand: 'Advanced Drainage Systems, Inc.',
            name: '120 x 7.5, 100 Sq Yd, Geosynthetic, Double, Photodegradable, Erosion Control Blanket',
            partNumber: '8X1125STRAWDN',
            mfr: '00S2TT',
            thumb: 'null',
            price: '43.59',
            qty: {
              quantityOrdered: 1,
              quantityShipped: 1
            }
          }
        ]
      }
    }
  }
};

export const successMincron_5448540 = {
  request: {
    query: GetInvoiceDocument,
    variables: {
      accountId: '91492',
      invoiceNumber: '5448540'
    }
  },
  result: {
    data: {
      invoice: {
        invoiceNumber: '5282411',
        status: 'Open',
        terms: 'NET 30 DAYS',
        customerPo: 'WILKESBORO JIM',
        invoiceDate: '10/06/2021',
        dueDate: '10/28/2020',
        originalAmt: '',
        openBalance: '',
        age: '',
        jobNumber: 'SHOP',
        jobName: 'SHOP',
        address: {
          streetLineOne: '911 BLUEBERRY WAY',
          streetLineTwo: '',
          streetLineThree: '',
          city: 'NORTHLAKE',
          state: 'TX',
          postalCode: '76226',
          country: ''
        },
        shipDate: '11/06/2021',
        deliveryMethod: 'Our Truck',
        subtotal: '2184.28',
        tax: '162.73',
        otherCharges: '0.0',
        paidToDate: '139.91',
        invoiceItems: {
          id: 'F02FD85F-4DCE-4EA7-8459-2DA7CFD689DC',
          brand: 'Certex USA, name:2 x 12, Nylon, Eye and Eye, Ply Sling',
          partNumber: 'SLING212,mfr:CX07-0126-12',
          thumb:
            'https://images.tradeservice.com/ProductImages/DIR100193/CERTEX_CX07-0126-12_LINE_SML.jpg',
          price: '40.00',
          qty: {
            quantityOrdered: 1,
            quantityShipped: 1
          }
        }
      }
    }
  }
};
