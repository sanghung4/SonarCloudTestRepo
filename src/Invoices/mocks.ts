import { subMonths } from 'date-fns';
import { Invoice, InvoicesDocument } from 'generated/graphql';
import { mockRow } from 'test-utils/mockTableInstance';
import { formatDate } from 'utils/dates';

export const success = {
  request: {
    query: InvoicesDocument,
    variables: {
      accountId: '123456',
      erpName: 'ECLIPSE'
    }
  },
  result: {
    data: {
      invoices: {
        bucketNinety: 0.0,
        bucketOneTwenty: 0.0,
        bucketSixty: 0.0,
        bucketThirty: 0.0,
        bucketFuture: 0.0,
        currentAmt: 35224.96,
        invoices: [
          {
            age: 'Future',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110227972.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobName: null,
            jobNumber: null
          },
          {
            age: 'Current',
            customerPo: '1096872',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110227975.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobName: null,
            jobNumber: null
          },
          {
            age: 'Over120',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110220294.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobName: null,
            jobNumber: null
          },
          {
            age: '91-120',
            customerPo: '109687',
            invoiceDate: '03/01/2021',
            dueDate: '03/01/2022',
            invoiceNumber: 'S110222345.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobName: null,
            jobNumber: null
          },
          {
            age: 'Deposit',
            customerPo: '109687',
            invoiceDate: '03/01/2021',
            dueDate: '03/01/2022',
            invoiceNumber: 'S110220974.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobName: null,
            jobNumber: null
          },
          {
            age: '31-60',
            customerPo: '109687',
            invoiceDate: '03/01/2021',
            dueDate: '03/01/2022',
            invoiceNumber: 'S110227973.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobName: null,
            jobNumber: null
          }
        ],
        totalAmt: 35232.27,
        totalPastDue: 7.3085938
      }
    }
  }
};

export const successMincron = {
  request: {
    query: InvoicesDocument,
    variables: {
      accountId: '123456',
      erpName: 'MINCRON'
    }
  },
  result: {
    data: {
      invoices: {
        bucketNinety: 0.0,
        bucketOneTwenty: 0.0,
        bucketSixty: 0.0,
        bucketThirty: 0.0,
        bucketFuture: 0.0,
        currentAmt: 35224.96,
        invoices: [
          {
            age: 'Future',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110227972.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobNumber: '6566',
            jobName: 'NEP 7 ASHVELLE'
          },
          {
            age: 'Current',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110227975.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobNumber: '56787',
            jobName: 'WALTER HORN'
          },
          {
            age: 'Over120',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110220294.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Closed',
            invoiceUrl: 'test.com',
            jobNumber: 'SHOP',
            jobName: 'SEQUAYESH TR'
          },
          {
            age: '91-120',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110222345.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobNumber: 'RETREAT',
            jobName: 'RETREAT AT ARDEN'
          },
          {
            age: 'Deposit',
            customerPo: '109687',
            invoiceDate: formatDate(subMonths(new Date(), 1)),
            dueDate: '03/01/2022',
            invoiceNumber: 'S110220974.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Closed',
            invoiceUrl: 'test.com',
            jobNumber: 'AVLNEP7',
            jobName: 'WALTER HORN'
          },
          {
            age: '31-60',
            customerPo: '109687',
            invoiceDate: '03/01/2021',
            dueDate: '03/01/2022',
            invoiceNumber: 'S110227973.001',
            openBalance: '12.39',
            originalAmt: '12.39',
            status: 'Open',
            invoiceUrl: 'test.com',
            jobNumber: 'AIKENON',
            jobName: 'AIKEN AVL'
          }
        ],
        totalAmt: 35232.27,
        totalPastDue: 7.3085938
      }
    }
  }
};

export const mockTestTableMemo = {
  invoices: {
    bucketNinety: 0.0,
    bucketOneTwenty: 0.0,
    bucketSixty: 0.0,
    bucketThirty: 0.0,
    bucketFuture: 0.0,
    currentAmt: 35224.96,
    invoices: [
      {
        age: 'Future',
        customerPo: '109687',
        invoiceDate: '05/02/2022',
        dueDate: '03/01/2022',
        invoiceNumber: 'S110227972.001',
        openBalance: '12.39',
        originalAmt: '12.39',
        status: 'Open',
        invoiceUrl: 'test.com',
        jobNumber: '6566',
        jobName: 'NEP 7 ASHVELLE'
      },
      {
        age: 'Current',
        customerPo: '109687',
        invoiceDate: '03/01/2021',
        dueDate: '03/01/2022',
        invoiceNumber: 'S110227975.001',
        openBalance: '12.39',
        originalAmt: '12.39',
        status: 'Open',
        invoiceUrl: 'test.com',
        jobNumber: '56787',
        jobName: 'WALTER HORN'
      },
      {
        age: 'Over120',
        customerPo: '109687',
        invoiceDate: '03/01/2021',
        dueDate: '03/01/2022',
        invoiceNumber: 'S110220294.001',
        openBalance: '12.39',
        originalAmt: '12.39',
        status: 'Open',
        invoiceUrl: 'test.com',
        jobNumber: 'SHOP',
        jobName: 'SEQUAYESH TR'
      },
      {
        age: '91-120',
        customerPo: '109687',
        invoiceDate: '03/01/2021',
        dueDate: '03/01/2022',
        invoiceNumber: 'S110222345.001',
        openBalance: '12.39',
        originalAmt: '12.39',
        status: 'Open',
        invoiceUrl: 'test.com',
        jobNumber: 'RETREAT',
        jobName: 'RETREAT AT ARDEN'
      },
      {
        age: 'Deposit',
        customerPo: '109687',
        invoiceDate: '03/01/2021',
        dueDate: '03/01/2022',
        invoiceNumber: 'S110220974.001',
        openBalance: '12.39',
        originalAmt: '12.39',
        status: 'Open',
        invoiceUrl: 'test.com',
        jobNumber: 'AVLNEP7',
        jobName: 'WALTER HORN'
      },
      {
        age: '31-60',
        customerPo: '109687',
        invoiceDate: '03/01/2021',
        dueDate: '03/01/2022',
        invoiceNumber: 'S110227973.001',
        openBalance: '12.39',
        originalAmt: '12.39',
        status: 'Open',
        invoiceUrl: 'test.com',
        jobNumber: 'AIKENON',
        jobName: 'AIKEN AVL'
      }
    ],
    totalAmt: 35232.27,
    totalPastDue: 7.3085938
  }
};

const rowToMock = success.result.data.invoices.invoices[0];
export const mockInvoiceRow = mockRow<Invoice>(rowToMock, rowToMock, {});
