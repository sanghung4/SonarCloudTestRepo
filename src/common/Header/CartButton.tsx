import { useContext } from 'react';

import {
  Badge,
  Box,
  Button,
  IconButton,
  useScreenSize
} from '@dialexa/reece-component-library';
import { CircularProgress } from '@mui/material';
import { useLocation, useHistory } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import { AuthContext } from 'AuthProvider';

import { cartButtonStyles } from 'common/Header/util';
import { CartIcon } from 'icons';
import { useCartContext } from 'providers/CartProvider';

export default function CartButton() {
  /**
   * Custom hooks
   */
  const location = useLocation();
  const { isSmallScreen } = useScreenSize();
  const { t } = useTranslation();

  /**
   * Context
   */
  const { cartLoading, itemLoading } = useCartContext();

  /**
   * Callbacks
   */
  const goToCart = useGoToCart();

  /**
   * Render
   */
  if (location.pathname !== '/select-accounts') {
    return isSmallScreen ? (
      <Box position="relative" ml={3}>
        {!!(cartLoading || itemLoading) && (
          <cartButtonStyles.CartLoaderComponent>
            <CircularProgress size="2rem" />
          </cartButtonStyles.CartLoaderComponent>
        )}
        <IconButton
          onClick={goToCart}
          size="large"
          data-testid="go-to-cart-button-mobile"
        >
          <CartButtonContent />
        </IconButton>
      </Box>
    ) : (
      <Button
        data-testid="go-to-cart-button-desktop"
        color="gray"
        iconColor="primary"
        startIcon={<CartButtonContent />}
        variant="text"
        onClick={goToCart}
        sx={cartButtonStyles.cartButtonSx}
      >
        {t('common.cart')}
      </Button>
    );
  }
  return null;
}

export function CartButtonContent() {
  /**
   * Hooks / Consts
   */
  const { isSmallScreen } = useScreenSize();
  const { cartLoading, itemCount, itemLoading } = useCartContext();
  const invisible = !!(cartLoading || itemLoading || !itemCount);
  const isLoading = !!(!isSmallScreen && (cartLoading || itemLoading));

  /**
   * Render
   */
  return (
    <Badge
      showZero={false}
      badgeContent={itemCount}
      invisible={invisible}
      max={999}
    >
      <CartIcon height={20} width={20} />
      {isLoading && (
        <cartButtonStyles.CartLoaderComponent>
          <CircularProgress size="2rem" />
        </cartButtonStyles.CartLoaderComponent>
      )}
    </Badge>
  );
}

export function useGoToCart() {
  /**
   * Hooks
   */
  const history = useHistory();
  const { authState } = useContext(AuthContext);
  const { checkingOutWithQuote, clearQuote, contract } = useCartContext();

  /**
   * Output
   */
  return () => {
    if (checkingOutWithQuote) {
      clearQuote();
    }
    const showNavAlert = authState?.isAuthenticated && contract;
    history.push(
      authState?.isAuthenticated ? '/cart' : '/login',
      showNavAlert ? { canShowNavAlert: true } : {}
    );
  };
}
