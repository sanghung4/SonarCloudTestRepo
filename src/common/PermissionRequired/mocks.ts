import { UserContextType } from 'AuthProvider';
import { Permission } from 'common/PermissionRequired';

export const authConfigNoPerm = {
  authState: {
    isAuthenticated: true
  },
  profile: {
    isEmployee: true,
    isVerified: true
  } as UserContextType['profile'],
  activeFeatures: ['LISTS']
};
export const authConfig = {
  authState: {
    isAuthenticated: true
  },
  profile: {
    userId: 'test',
    permissions: [Permission.EDIT_LIST, Permission.APPROVE_CART],
    isEmployee: true,
    isVerified: true
  } as UserContextType['profile'],
  activeFeatures: ['LISTS']
};
