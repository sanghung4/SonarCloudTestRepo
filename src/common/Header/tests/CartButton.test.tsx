import { fireEvent } from '@testing-library/react';

import { mockAuthContext } from 'AuthProvider.mocks';
import { mockCartContext, mockCartContract } from 'Cart/tests/mocks';
import CartButton, { useGoToCart } from 'common/Header/CartButton';
import { setBreakpoint } from 'test-utils/mockMediaQuery';
import { render } from 'test-utils/TestWrapper';

let mockCart = { ...mockCartContext };
let mockAuth = { ...mockAuthContext };
let mockLocation = { pathname: '' };
let mockHistory = { push: jest.fn() };

jest.mock('react-router-dom', () => ({
  useLocation: () => mockLocation,
  useHistory: () => mockHistory
}));

function MockComponent() {
  const goToCart = useGoToCart();
  return <button data-testid="button" onClick={goToCart} />;
}

describe('common - header - CartButton', () => {
  afterEach(() => {
    mockCart = { ...mockCartContext };
    mockAuth = { ...mockAuthContext };
    mockLocation = { pathname: '' };
    mockHistory = { push: jest.fn() };
  });

  it('should match snapshot on desktop', () => {
    setBreakpoint('desktop');
    const { container } = render(<CartButton />);
    expect(container).toMatchSnapshot();
  });
  it('should match snapshot on mobile', () => {
    setBreakpoint('mobile');
    const { container } = render(<CartButton />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot while loading on desktop', () => {
    setBreakpoint('desktop');
    mockCart.cartLoading = true;
    mockCart.itemLoading = 'test';
    const { container } = render(<CartButton />, { cartConfig: mockCart });
    expect(container).toMatchSnapshot();
  });
  it('should match snapshot while loading on mobile', () => {
    setBreakpoint('mobile');
    mockCart.cartLoading = true;
    mockCart.itemLoading = 'test';
    const { container } = render(<CartButton />, { cartConfig: mockCart });
    expect(container).toMatchSnapshot();
  });

  it('should result in null when location is "/select-accounts"', () => {
    setBreakpoint('desktop');
    mockLocation.pathname = '/select-accounts';
    const { container } = render(<CartButton />);
    expect(container).toMatchSnapshot();
  });

  it('expect goToCart to call clear quote', () => {
    mockCart.checkingOutWithQuote = true;
    const { getByTestId } = render(<MockComponent />, {
      cartConfig: mockCart,
      authConfig: mockAuth
    });
    fireEvent.click(getByTestId('button'));
    expect(mockCart.clearQuote).toBeCalled();
    expect(mockHistory.push).toBeCalledWith('/login', {});
  });
  it('expect goToCart to call to go to cart', () => {
    mockAuth.authState = { isAuthenticated: true };
    const { getByTestId } = render(<MockComponent />, {
      cartConfig: mockCart,
      authConfig: mockAuth
    });
    fireEvent.click(getByTestId('button'));
    expect(mockHistory.push).toBeCalledWith('/cart', {});
  });
  it('expect goToCart to call to go to cart with contract', () => {
    mockAuth.authState = { isAuthenticated: true };
    mockCart.contract = mockCartContract;
    const { getByTestId } = render(<MockComponent />, {
      cartConfig: mockCart,
      authConfig: mockAuth
    });
    fireEvent.click(getByTestId('button'));
    expect(mockHistory.push).toBeCalledWith('/cart', { canShowNavAlert: true });
  });
});
