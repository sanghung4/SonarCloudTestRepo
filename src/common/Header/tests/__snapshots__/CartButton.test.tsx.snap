// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common - header - CartButton should match snapshot on desktop 1`] = `
<div>
  <button
    class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-1k4g8tg-MuiButtonBase-root-MuiButton-root"
    data-testid="go-to-cart-button-desktop"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-startIcon MuiButton-iconSizeMedium css-1sx3uk1-MuiButton-startIcon"
    >
      <span
        class="MuiBadge-root css-cei2pf-MuiBadge-root"
      >
        <svg
          height="20"
          width="20"
        >
          shopping-cart.svg
        </svg>
        <span
          class="MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-invisible css-13k9y7l-MuiBadge-badge"
        />
      </span>
    </span>
    Cart
  </button>
</div>
`;

exports[`common - header - CartButton should match snapshot on mobile 1`] = `
<div>
  <div
    class="MuiBox-root css-1tkfj2c"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-1unv2py-MuiButtonBase-root-MuiIconButton-root"
      data-testid="go-to-cart-button-mobile"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBadge-root css-cei2pf-MuiBadge-root"
      >
        <svg
          height="20"
          width="20"
        >
          shopping-cart.svg
        </svg>
        <span
          class="MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-invisible css-13k9y7l-MuiBadge-badge"
        >
          0
        </span>
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`common - header - CartButton should match snapshot while loading on desktop 1`] = `
<div>
  <button
    class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-1k4g8tg-MuiButtonBase-root-MuiButton-root"
    data-testid="go-to-cart-button-desktop"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-startIcon MuiButton-iconSizeMedium css-1sx3uk1-MuiButton-startIcon"
    >
      <span
        class="MuiBadge-root css-cei2pf-MuiBadge-root"
      >
        <svg
          height="20"
          width="20"
        >
          shopping-cart.svg
        </svg>
        <div
          class="MuiBox-root css-2go2wp"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-p36nji-MuiCircularProgress-root"
            role="progressbar"
            style="width: 2rem; height: 2rem;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
        </div>
        <span
          class="MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-invisible css-13k9y7l-MuiBadge-badge"
        />
      </span>
    </span>
    Cart
  </button>
</div>
`;

exports[`common - header - CartButton should match snapshot while loading on mobile 1`] = `
<div>
  <div
    class="MuiBox-root css-1tkfj2c"
  >
    <div
      class="MuiBox-root css-2go2wp"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-p36nji-MuiCircularProgress-root"
        role="progressbar"
        style="width: 2rem; height: 2rem;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-1unv2py-MuiButtonBase-root-MuiIconButton-root"
      data-testid="go-to-cart-button-mobile"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBadge-root css-cei2pf-MuiBadge-root"
      >
        <svg
          height="20"
          width="20"
        >
          shopping-cart.svg
        </svg>
        <span
          class="MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-invisible css-13k9y7l-MuiBadge-badge"
        >
          0
        </span>
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`common - header - CartButton should result in null when location is "/select-accounts" 1`] = `<div />`;
